reader_compare.go

```diff
@@ -0,0 +1,419 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package ionhash
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+
+	"github.com/amzn/ion-go/ion"
+)
+
+func Compare(reader1 ion.Reader, reader2 ion.Reader) error {
+	for true {
+		next, err := HasNext(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		if !next {
+			break
+		}
+
+		ionType1 := reader1.Type()
+		ionType2 := reader2.Type()
+
+		if ionType1 != ionType2 {
+			return fmt.Errorf("expected ion types to match")
+		}
+
+		ionHashReader, ok := reader1.(*hashReader)
+		if ok && ionHashReader.isInStruct() {
+			err := CompareFieldNames(reader1, reader2)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = CompareAnnotations(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		err = CompareAnnotationSymbols(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		err = CompareHasAnnotations(reader1, reader2)
+		if err != nil {
+			return err
+		}
+
+		isNull1 := reader1.IsNull()
+		isNull2 := reader2.IsNull()
+		if isNull1 != isNull2 {
+			return fmt.Errorf("expected readers to have same value for IsNull()")
+		}
+
+		switch ionType1 {
+		case ion.NullType:
+			if !isNull1 {
+				return fmt.Errorf("expected ionType1 to be null")
+			}
+			if !isNull2 {
+				return fmt.Errorf("expected ionType2 to be null")
+			}
+			break
+		case ion.BoolType, ion.IntType, ion.FloatType, ion.DecimalType, ion.TimestampType,
+			ion.StringType, ion.SymbolType, ion.BlobType, ion.ClobType:
+			err := CompareScalars(ionType1, isNull1, reader1, reader2)
+			if err != nil {
+				return err
+			}
+			break
+		case ion.StructType, ion.ListType, ion.SexpType:
+			err := reader1.StepIn()
+			if err != nil {
+				return err
+			}
+
+			err = reader2.StepIn()
+			if err != nil {
+				return err
+			}
+
+			err = Compare(reader1, reader2)
+			if err != nil {
+				return err
+			}
+
+			err = reader1.StepOut()
+			if err != nil {
+				return err
+			}
+
+			err = reader2.StepOut()
+			if err != nil {
+				return err
+			}
+			break
+		default:
+			return &InvalidIonTypeError{ionType1}
+		}
+	}
+
+	next, err := HasNext(reader1, reader2)
+	if err != nil {
+		return err
+	}
+
+	if next {
+		return fmt.Errorf("expected HasNext() to return false")
+	}
+
+	return nil
+}
+
+func HasNext(reader1 ion.Reader, reader2 ion.Reader) (bool, error) {
+	more := reader1.Next()
+
+	if more != reader2.Next() {
+		return false, fmt.Errorf("next results don't match")
+	}
+
+	if !more {
+		if !reader1.Next() {
```

Comment:
this will call `Next()` a second time on `reader1` who already returned `false`. This seems odd to me. Can you explain why we need the second call to `Next()` after the immediately preceding call to `Next()` returned `false`?

Reply:
It is to assert that the second `Next()` call will still return false

Reply:
Is that an assertion that we should be relying on? I do not recall the Reader's API saying anything about the result of a second call to `Next()` if the first call returned `false`.

Reply:
I will change it to check the Err() instead of a second call to Next().

Reply:
I think the assertion is appropriate because we are testing that advancing the reader to the next position which doesn't exist will return false. I think we should also add an Err is nil check to ensure that it wasn't due to an error.

API documentation:

> 	// Next advances the Reader to the next position in the current value stream.
> 	// It returns true if this is the position of an Ion value, and false if it
> 	// is not. On error, it returns false and sets Err.
>

Reply:
I see. In that case, if `HasNext` is supposed to return true when there is something next in the stream (not the end of the stream) then checking that we got no `Err()` in the reader is all we need. But if `HasNext()` is also supposed to check that we got something valid (but perhaps not the same value) then we need to check the booleans returned from `Next` are both `true`.

Can we add a comment stating the intent of this function and then update the implementation to meet that intent.

Reply:
Addressed.